File: lib/services/auth_service.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import '../models/user.dart';
import 'package:http/http.dart' as http;

class AuthService {
  static const String _userKey = 'user';
  static const String _tokenKey = 'token';
  static User? _currentUser;
  static const String baseUrl = 'http://localhost:3000/api';
  static const String coreIoTBaseUrl = 'https://app.coreiot.io/api';
  static const String coreIoTUsername = 'an.nguyencse03@gmail.com';
  static const String coreIOTPassword = '02121209An';

  // Get the current logged-in user
  static User? get currentUser => _currentUser;

  // Initialize the auth service
  static Future<void> init() async {
    final prefs = await SharedPreferences.getInstance();
    final userJson = prefs.getString(_userKey);
    if (userJson != null) {
      _currentUser = User.fromJson(json.decode(userJson));
      // Try to refresh CoreIoT token if user exists
      await loginCoreIoT();
    }
  }

  // Login to Core IoT
  static Future<String?> loginCoreIoT() async {
    try {
      final response = await http.post(
        Uri.parse('$coreIoTBaseUrl/auth/login'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode({
          'username': coreIoTUsername,
          'password': coreIOTPassword,
        }),
      );

      print('CoreIoT Status: ${response.statusCode}');
      print('CoreIoT Response: ${response.body}');

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        final token = data['token'];
        // Save token to SharedPreferences
        final prefs = await SharedPreferences.getInstance();
        await prefs.setString(_tokenKey, token);
        return token;
      }
      return null;
    } catch (e) {
      print('CoreIoT Login error: $e');
      return null;
    }
  }

  // Login with username and password
  static Future<User> login(String username, String password) async {
    try {
      final token = await loginCoreIoT();
      if (token == null) {
        throw Exception('Failed to login to CoreIoT');
      }

      final user = User(
        id: 1,
        username: username,
        role: 'user',
        createdAt: DateTime.now(),
      );

      final prefs = await SharedPreferences.getInstance();
      await prefs.setString(_userKey, json.encode(user.toJson()));

      _currentUser = user;
      return user;
    } catch (e) {
      print('Login error: $e');
      throw Exception('Failed to login: $e');
    }
  }

  // Register a new user
  static Future<User> register(String username, String password, String role) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/auth/register'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode({
          'username': username,
          'password': password,
          'role': role,
        }),
      );

      if (response.statusCode == 201) {
        final data = json.decode(response.body);
        final user = User(
          id: data['user']['id'],
          username: data['user']['username'],
          role: data['user']['role'],
          createdAt: DateTime.parse(data['user']['createdAt'] ?? DateTime.now().toIso8601String()),
        );

        // Save user to shared preferences
        final prefs = await SharedPreferences.getInstance();
        await prefs.setString(_userKey, json.encode(user.toJson()));

        _currentUser = user;

        // Call Core IoT API after successful registration
        await loginCoreIoT();

        return user;
      } else {
        throw Exception(json.decode(response.body)['message'] ?? 'Failed to register');
      }
    } catch (e) {
      print('Registration error: $e');
      throw Exception('Failed to register: $e');
    }
  }

  // Logout the current user
  static Future<void> logout() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_userKey);
    await prefs.remove(_tokenKey);
    _currentUser = null;
  }

  // Check if user is logged in
  static bool get isLoggedIn => _currentUser != null;

  // Get current token
  static Future<String?> getToken() async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString(_tokenKey);
    if (token == null) {
      // Try to refresh token if not found
      return await loginCoreIoT();
    }
    return token;
  }
}

File: src/services/authService.js
const jwt = require('jsonwebtoken');
const bcrypt = require('bcryptjs');
const database = require('../config/database');
const axios = require('axios');

class AuthService {
  // Generate JWT token
  static generateToken(user) {
    return jwt.sign(
      {
        userId: user.id,
        username: user.username,
        role: user.role
      },
      process.env.JWT_SECRET,
      { expiresIn: '24h' }
    );
  }

  // Verify JWT token
  static verifyToken(token) {
    try {
      return jwt.verify(token, process.env.JWT_SECRET);
    } catch (error) {
      throw new Error('Invalid token');
    }
  }

  // Register new user
  static async register(userData) {
    const { username, password, role } = userData;
    const pool = await database.connect();

    // Check if username exists
    const existingUser = await pool.request()
      .input('username', database.sql.NVarChar, username)
      .query('SELECT id FROM Users WHERE username = @username');

    if (existingUser.recordset.length > 0) {
      throw new Error('Username already exists');
    }

    // Hash password
    const hashedPassword = await bcrypt.hash(password, 10);

    // Create user
    const result = await pool.request()
      .input('username', database.sql.NVarChar, username)
      .input('password', database.sql.NVarChar, hashedPassword)
      .input('role', database.sql.NVarChar, role)
      .query(`
        INSERT INTO Users (username, password, role)
        VALUES (@username, @password, @role)
        SELECT SCOPE_IDENTITY() as id
      `);

    const userId = result.recordset[0].id;

    // Generate token
    const token = this.generateToken({ id: userId, username, role });

    return {
      token,
      user: {
        id: userId,
        username,
        role
      }
    };
  }

  // Login user
  static async login(credentials) {
    const { username, password } = credentials;
    const pool = await database.connect();

    try {
      // Get user
      const result = await pool.request()
        .input('username', database.sql.NVarChar, username)
        .query('SELECT * FROM Users WHERE username = @username');

      const user = result.recordset[0];

      if (!user) {
        throw new Error('Invalid credentials');
      }

      // Verify password
      const isValidPassword = await bcrypt.compare(password, user.password);
      if (!isValidPassword) {
        throw new Error('Invalid credentials');
      }

      // Generate token
      const token = this.generateToken(user);

      return {
        token,
        user: {
          id: user.id,
          username: user.username,
          role: user.role
        }
      };
    } catch (error) {
      console.error('Login error:', error);
      throw error;
    }
  }

  // Refresh token
  static async refreshToken(oldToken) {
    try {
      const decoded = this.verifyToken(oldToken);
      const pool = await database.connect();

      // Get user
      const result = await pool.request()
        .input('userId', database.sql.Int, decoded.userId)
        .query('SELECT * FROM Users WHERE id = @userId');

      const user = result.recordset[0];

      if (!user) {
        throw new Error('User not found');
      }

      // Generate new token
      const token = this.generateToken(user);

      return {
        token,
        user: {
          id: user.id,
          username: user.username,
          role: user.role
        }
      };
    } catch (error) {
      throw new Error('Invalid token');
    }
  }

  static coreIoTBaseUrl = 'https://app.coreiot.io/api';
  static coreIoTUsername = 'an.nguyencse03@gmail.com';
  static coreIoTPassword = '02121209An';

  // Login to Core IoT
  static async loginCoreIoT() {
    try {
      const response = await axios.post(
        `${this.coreIoTBaseUrl}/auth/login`,
        {
          username: this.coreIoTUsername,
          password: this.coreIoTPassword,
        },
        {
          headers: {'Content-Type': 'application/json'}
        }
      );

      console.log('CoreIoT Login Status:', response.status);
      console.log('CoreIoT Response:', response.data);

      if (response.status === 200) {
        return response.data.token;
      }
      return null;
    } catch (error) {
      console.error('CoreIoT Login error:', error.response?.data || error.message);
      return null;
    }
  }
}

module.exports = AuthService;